File: packages/webservice/models/src/main/kotlin/com/yahoo/navi/ws/models/beans/fragments/request/Sort.kt
/**
 * Copyright 2020, Yahoo Holdings Inc.
 * Licensed under the terms of the MIT license. See accompanying LICENSE.md file for terms.
 */
package com.yahoo.navi.ws.models.beans.fragments.request

import com.yahoo.navi.ws.models.beans.enums.ColumnarType
import org.hibernate.annotations.Parameter
import org.hibernate.annotations.Type

data class Sort(
    var field: String = "",
    @Type(
        type = "com.yahoo.navi.ws.models.types.JsonType",
        parameters = [
            Parameter(name = "class", value = "java.utils.HashMap")
        ]
    ) var parameters: Map<String, String> = emptyMap(),
    var type: ColumnarType? = null,
    var direction: String = ""
)

File: packages/core/addon/models/request/sort.ts
/**
 * Copyright 2022, Yahoo Holdings Inc.
 * Licensed under the terms of the MIT license. See accompanying LICENSE.md file for terms.
 */
import { attr } from '@ember-data/model';
import { validator, buildValidations } from 'ember-cp-validations';
import BaseFragment from '../request/base';
import type { SortDirection, Sort } from '@yavin/client/request';

const Validations = buildValidations({
  direction: validator('inclusion', {
    in: ['asc', 'desc'],
    message: 'The `direction` sort field must equal to `asc` or `desc`',
  }),
});

/**
 * @augments {BaseFragment}
 */
export default class SortFragment extends BaseFragment.extend(Validations) implements Sort {
  @attr('string', { defaultValue: 'desc' })
  declare direction: SortDirection;

  @attr()
  declare cid: string;
}

declare module 'navi-core/models/registry' {
  export interface FragmentRegistry {
    'request/sort': SortFragment;
  }
}
